app/services/sso/openid_connect.rb

```diff
@@ -4,21 +4,40 @@ class OpenidConnect < Base
     attr_reader :current_user
 
     def available?
+      if session.present? && (session[:sso_method] == "SSO::OpenidConnect")
+        session.delete(:sso_method)
+        return true
+      end
       ((controller.api_request? && bearer_token_set?) || http_token.present?) && valid_issuer?
     end
 
-    def authenticate!
-      payload = jwt_token.decode
-      return nil if payload.nil?
-      user = find_or_create_user_from_jwt(payload)
+    def authenticated?
+      return false unless (payload = jwt_token.decode)
+      return false unless (user = find_or_create_user_from_jwt(payload))
+
       @current_user = user
       update_session(payload)
-      user&.login
+      true
     end
 
-    def authenticated?
-      self.user = User.current.presence || authenticate!
+    def authenticate!
+      self.has_rendered = true
+      controller.redirect_to login_url
```

Comment:
Shouldn't the redirect be to the extlogin path?

Reply:
When we use `extlogin_users_path`, the browser picks up the previous access token and tries to login with the expired token. Instead, of asking for a new token from the Openid Provider.

Therefore, browser redirects to the Foreman login page.

To be more clear, if I do something like this
```
def login_url
  controller.main_app.extlogin_users_path
end

def authenticate!
  self.has_rendered = true
  controller.redirect_to login_url
end
```

Then, this is the log for the above code.
```
Started GET "/users/extlogin" for 192.168.122.1 at 2020-04-07 10:55:12 +0000
[I|app|ea777322] Processing by UsersController#extlogin as HTML
[W|app|ea777322] Failed to decode JWT
[D|app|ea777322] Backtrace for 'Failed to decode JWT' error (JWT::ExpiredSignature): Signature has expired
/home/vagrant/foreman/app/services/sso/openid_connect.rb:15:in `authenticated?'
/home/vagrant/foreman/app/controllers/concerns/foreman/controller/authentication.rb:55:in `sso_authentication'
``` 

When we call[ authenticate!](https://github.com/theforeman/foreman/blob/develop/app/controllers/concerns/foreman/controller/authentication.rb#L59) it means that there is no user in the session or current user is not set. see here https://github.com/theforeman/foreman/blob/develop/app/controllers/concerns/foreman/controller/authentication.rb#L9

In the case of `mod_auth_oidc`, we need to perform the logout if we want re-login, since it stores values in the cookies.
 
When we use `expiration_url`, the session is cleared and the user makes a fresh request to the OpenID provider and therefore is redirected to the OpenID provider page instead of the Foreman login page.

@tbrisker let me know your thoughts.